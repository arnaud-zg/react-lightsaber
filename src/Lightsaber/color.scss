.lightsaber-component {
  .lightsaber-component__blade {
    // PEACE THROUGH FORCE
    &.peace {
      color: #0c9655;
    }

    // EVIL AND POWER
    &.evil {
      color: #ca1b21;
    }

    // JUSTICE AND PROTECTION
    &.justice {
      color: #52b6e0;
    }

    // MORAL AMBIGUITY
    &.moral {
      color: #bd80bd;
    }

    // INTRIGUE AND PURSUIT
    &.intrigue {
      color: #fff8c9;
    }

    // NEGOCIATION
    &.negocition {
      color: #bb4923;
    }

    // STRENGTH
    &.strength {
      color: #f7df41;
    }

    // PHYSICAL STRENGTH
    &.physical {
      color: #bd7a3a;
    }

    // CALM AND CLEAR
    &.calm {
      color: #bbbbbd;
    }

    // SELF OBSESSION
    &.self {
      color: #1c1c1e;
    }

    // OBEDIENCE AND SERVICE
    &.obedience {
      color: #ffffff;
    }
  }
}
