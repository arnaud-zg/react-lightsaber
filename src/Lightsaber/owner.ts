export enum ELightsaberOwners {
  ANAKIN_SKYWALKER = 'ANAKIN_SKYWALKER',
  DARTH_MAUL = 'DARTH_MAUL',
  DARTH_SIDIOUS = 'DARTH_SIDIOUS',
  DARTH_VADER = 'DARTH_VADER',
  KYLO_REN = 'KYLO_REN',
  LOWBACCA = 'LOWBACCA',
  LUKE_SKYWALKER = 'LUKE_SKYWALKER',
  MACE_WINDU = 'MACE_WINDU',
  MARA_JADE_SKYWALKER = 'MARA_JADE_SKYWALKER',
  OBI_WAN = 'OBI_WAN',
  PLO_KOON = 'PLO_KOON',
  PRE_VIZSLA = 'PRE_VIZSLA',
  QIU_GON_JINN = 'QIU_GON_JINN',
  QU_RAHN = 'QU_RAHN',
  TERA_SINUBE = 'TERA_SINUBE',
  YADDLE = 'YADDLE',
  YARAEL_POOF = 'YARAEL_POOF',
  YODA = 'YODA',
}

export enum ELightsaberColorClassname {
  CALM = 'calm',
  EVIL = 'evil',
  INTRIGUE = 'intrigue',
  JUSTICE = 'justice',
  MORAL = 'moral',
  NEGOCIATION = 'negociation',
  OBEDIENCE = 'obedience',
  PEACE = 'peace',
  PHYSICAL = 'physical',
  SELF = 'self',
  STRENGTH = 'strength',
}

export const LIGHTSABER_COLOR_CLASSNAME_MAPPING = {
  [ELightsaberOwners.ANAKIN_SKYWALKER]: ELightsaberColorClassname.JUSTICE,
  [ELightsaberOwners.DARTH_MAUL]: ELightsaberColorClassname.EVIL,
  [ELightsaberOwners.DARTH_SIDIOUS]: ELightsaberColorClassname.EVIL,
  [ELightsaberOwners.DARTH_VADER]: ELightsaberColorClassname.EVIL,
  [ELightsaberOwners.KYLO_REN]: ELightsaberColorClassname.EVIL,
  [ELightsaberOwners.LOWBACCA]: ELightsaberColorClassname.PHYSICAL,
  [ELightsaberOwners.LUKE_SKYWALKER]: ELightsaberColorClassname.PEACE,
  [ELightsaberOwners.MACE_WINDU]: ELightsaberColorClassname.MORAL,
  [ELightsaberOwners.MARA_JADE_SKYWALKER]: ELightsaberColorClassname.MORAL,
  [ELightsaberOwners.OBI_WAN]: ELightsaberColorClassname.JUSTICE,
  [ELightsaberOwners.PLO_KOON]: ELightsaberColorClassname.INTRIGUE,
  [ELightsaberOwners.PRE_VIZSLA]: ELightsaberColorClassname.SELF,
  [ELightsaberOwners.QIU_GON_JINN]: ELightsaberColorClassname.PEACE,
  [ELightsaberOwners.QU_RAHN]: ELightsaberColorClassname.STRENGTH,
  [ELightsaberOwners.TERA_SINUBE]: ELightsaberColorClassname.CALM,
  [ELightsaberOwners.YADDLE]: ELightsaberColorClassname.NEGOCIATION,
  [ELightsaberOwners.YARAEL_POOF]: ELightsaberColorClassname.INTRIGUE,
  [ELightsaberOwners.YODA]: ELightsaberColorClassname.PEACE,
} as { [key: string]: string }
