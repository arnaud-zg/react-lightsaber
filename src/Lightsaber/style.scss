@import '../_common/variable.scss';

.lightsaber-component {
  display: flex;
  flex-direction: row;
  font-family: Courier, monospace;
  font-size: 100%;
  margin: 10px 0;
  position: relative;
  min-width: 330px;

  input[type='checkbox'] {
    display: none;

    &:not(:checked) ~ .lightsaber-component__blade {
      flex: 5;

      &:after {
        transition-duration: 0.3s;
        width: 0%;
      }
    }

    &:checked {
      ~ .lightsaber-component__controller {
        .lightsaber-component__switch {
          &:before {
            color: $color-3;
          }
        }
      }
    }
  }

  .lightsaber-component__blade {
    flex: 5;
    position: relative;
    z-index: $layer-level-2;

    &:after,
    &:before {
      content: '';
      position: absolute;
      transition-property: width;
    }

    &:before {
      background-color: $color-5;
      border-radius: 0 10px 10px 0;
      display: block;
      filter: blur(1px);
      height: 50%;
      top: 25%;
      transition-duration: 0.15s;
      width: 98%;
      z-index: $layer-level-3;
    }

    &:after {
      background-color: currentColor;
      border-radius: 0 10px 10px 0;
      display: block;
      filter: blur(4px);
      height: 100%;
      transition-duration: 0.3s;
      width: 100%;
      z-index: $layer-level-2;
    }
  }

  .lightsaber-component__controller {
    background-color: $color-2;
    color: $color-4;
    cursor: pointer;
    flex: 1;
    position: relative;
    user-select: none;
    z-index: $layer-level-3;
    max-width: 15%;

    .lightsaber-component__switch {
      color: $color-1;
      content: '¦¦¦¦';
      font-size: 150%;
      font-weight: bold;
      letter-spacing: -0.65rem;
      padding-left: 25%;

      &:after,
      &:before {
        height: 0;
        position: absolute;
      }

      &:before {
        color: $color-4;
        content: '●';
        font-size: 130%;
        font-weight: bold;
        left: -2%;
        top: -10%;
        z-index: $layer-level-3;
      }

      &:after {
        color: $color-2;
        content: '◀';
        font-size: 250%;
        right: 0;
        top: -95%;
        z-index: $layer-level-2;
      }
    }

    &:after,
    &:before {
      content: '';
      display: block;
      position: absolute;
      z-index: $layer-level-3;
    }

    &:before {
      background-color: $color-1;
      border-radius: 10px 0 0 10px;
      height: 90%;
      left: -10%;
      top: 5%;
      width: 10%;
    }

    &:after {
      background-color: $color-2;
      height: 80%;
      right: -30%;
      top: 10%;
      width: 20%;
    }
  }
}
