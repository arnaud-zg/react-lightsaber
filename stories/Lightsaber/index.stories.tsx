import React from 'react'
import { storiesOf } from '@storybook/react'
import Lightsaber from '../../src/Lightsaber'
import { ELightsaberOwners } from '../../src/Lightsaber/owner'

Object.keys(ELightsaberOwners).map((ownerKey) => {
  const owner = ELightsaberOwners[ownerKey as ELightsaberOwners]

  storiesOf('Lightsaber', module).add(`of ${owner}`, () => (
    <>
      <h2>Simple</h2>

      <div style={{ width: '360px' }}>
        <Lightsaber id="1" owner={owner} />
      </div>

      <div style={{ width: '360px' }}>
        <Lightsaber id="2" owner={owner} initialChecked={true} />
      </div>
    </>
  ))
})
