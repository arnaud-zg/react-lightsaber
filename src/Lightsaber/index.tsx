import React from 'react'
import './style.scss'

export enum ELightsaberOwners {
  YODA = 'yoda',
  VADER = 'vader',
  WINDU = 'windu',
  OBI = 'obi-wan',
}

interface IProps {
  id: string
  owner: ELightsaberOwners
}

const Lightsaber = (props: IProps) => {
  const { id, owner } = props

  return (
    <div className="lightsaber-component">
      <input type="checkbox" id={id} className={`jedi jedi--${owner}`} />
    </div>
  )
}

export default Lightsaber
