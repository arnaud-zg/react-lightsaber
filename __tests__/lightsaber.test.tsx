import React from 'react'
import ReactRenderer from 'react-test-renderer'
import Lightsaber from '../src/Lightsaber'
import { ELightsaberOwners } from '../src/Lightsaber/owner'

describe('Component.Lightsaber()', () => {
  it('should render component', () => {
    const component = ReactRenderer.create(
      <Lightsaber id="1" owner={ELightsaberOwners.OBI_WAN} />,
    )

    let tree = component.toJSON()

    expect(tree).toBeTruthy()
    expect(tree).toMatchSnapshot()
  })
})
