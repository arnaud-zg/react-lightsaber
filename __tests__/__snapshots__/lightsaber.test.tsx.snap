// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component.Lightsaber() should render component 1`] = `
<div
  className="lightsaber-component"
>
  <input
    checked={false}
    id="1"
    onClick={[Function]}
    type="checkbox"
  />
  <label
    className="lightsaber-component__controller"
    htmlFor="1"
  >
    <span
      className="lightsaber-component__switch"
    >
      ¦¦¦¦
    </span>
  </label>
  <span
    className="lightsaber-component__blade justice"
  />
</div>
`;
